File: client/src-tauri/src/hwid.rs
use once_cell::sync::Lazy;
use std::process::Command;

static HWID: Lazy<String> = Lazy::new(|| generate_hwid());

pub fn get_hwid() -> &'static str {
    &HWID
}

fn generate_hwid() -> String {
    let mut components = Vec::new();

    // Get CPU ID
    if let Some(cpu_id) = get_cpu_id() {
        components.push(cpu_id);
    }

    // Get disk serial
    if let Some(disk_serial) = get_disk_serial() {
        components.push(disk_serial);
    }

    // Get MAC address
    if let Some(mac) = get_mac_address() {
        components.push(mac);
    }

    // Get motherboard serial
    if let Some(mb_serial) = get_motherboard_serial() {
        components.push(mb_serial);
    }

    // If no components found, generate a fallback ID
    if components.is_empty() {
        components.push(generate_fallback_id());
    }

    // Combine all components and hash them
    let combined = components.join("|");
    hash_string(&combined)
}

fn get_cpu_id() -> Option<String> {
    #[cfg(target_os = "windows")]
    {
        let output = Command::new("wmic")
            .args(["cpu", "get", "ProcessorId"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines().skip(1) {
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                return Some(trimmed.to_string());
            }
        }
    }

    #[cfg(target_os = "linux")]
    {
        if let Ok(content) = std::fs::read_to_string("/proc/cpuinfo") {
            for line in content.lines() {
                if line.starts_with("Serial") || line.contains("model name") {
                    if let Some(value) = line.split(':').nth(1) {
                        return Some(value.trim().to_string());
                    }
                }
            }
        }
    }

    #[cfg(target_os = "macos")]
    {
        let output = Command::new("sysctl")
            .args(["-n", "machdep.cpu.brand_string"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        let trimmed = text.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }

    #[cfg(target_os = "android")]
    {
        // Android: try to get Build.SERIAL or ANDROID_ID
        let output = Command::new("getprop")
            .args(["ro.serialno"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        let trimmed = text.trim();
        if !trimmed.is_empty() && trimmed != "unknown" {
            return Some(trimmed.to_string());
        }
    }

    None
}

fn get_disk_serial() -> Option<String> {
    #[cfg(target_os = "windows")]
    {
        let output = Command::new("wmic")
            .args(["diskdrive", "get", "SerialNumber"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines().skip(1) {
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                return Some(trimmed.to_string());
            }
        }
    }

    #[cfg(target_os = "linux")]
    {
        // Try to get disk ID from /dev/disk/by-id
        if let Ok(entries) = std::fs::read_dir("/dev/disk/by-id") {
            for entry in entries.flatten() {
                let name = entry.file_name().to_string_lossy().to_string();
                if name.starts_with("ata-") || name.starts_with("nvme-") {
                    return Some(name);
                }
            }
        }
    }

    #[cfg(target_os = "macos")]
    {
        let output = Command::new("system_profiler")
            .args(["SPSerialATADataType"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines() {
            if line.contains("Serial Number") {
                if let Some(value) = line.split(':').nth(1) {
                    return Some(value.trim().to_string());
                }
            }
        }
    }

    None
}

fn get_mac_address() -> Option<String> {
    #[cfg(target_os = "windows")]
    {
        let output = Command::new("getmac")
            .args(["/fo", "csv", "/nh"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines() {
            let parts: Vec<&str> = line.split(',').collect();
            if let Some(mac) = parts.first() {
                let mac = mac.trim().trim_matches('"');
                if !mac.is_empty() && mac.contains('-') {
                    return Some(mac.to_string());
                }
            }
        }
    }

    #[cfg(any(target_os = "linux", target_os = "android"))]
    {
        if let Ok(entries) = std::fs::read_dir("/sys/class/net") {
            for entry in entries.flatten() {
                let name = entry.file_name().to_string_lossy().to_string();
                if name == "lo" {
                    continue;
                }
                let addr_path = entry.path().join("address");
                if let Ok(mac) = std::fs::read_to_string(addr_path) {
                    let mac = mac.trim();
                    if !mac.is_empty() && mac != "00:00:00:00:00:00" {
                        return Some(mac.to_string());
                    }
                }
            }
        }
    }

    #[cfg(target_os = "macos")]
    {
        let output = Command::new("ifconfig")
            .args(["en0"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines() {
            if line.contains("ether") {
                if let Some(mac) = line.split_whitespace().nth(1) {
                    return Some(mac.to_string());
                }
            }
        }
    }

    None
}

fn get_motherboard_serial() -> Option<String> {
    #[cfg(target_os = "windows")]
    {
        let output = Command::new("wmic")
            .args(["baseboard", "get", "SerialNumber"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines().skip(1) {
            let trimmed = line.trim();
            if !trimmed.is_empty() && trimmed != "To be filled by O.E.M." {
                return Some(trimmed.to_string());
            }
        }
    }

    #[cfg(target_os = "linux")]
    {
        if let Ok(serial) = std::fs::read_to_string("/sys/class/dmi/id/board_serial") {
            let trimmed = serial.trim();
            if !trimmed.is_empty() {
                return Some(trimmed.to_string());
            }
        }
    }

    #[cfg(target_os = "macos")]
    {
        let output = Command::new("ioreg")
            .args(["-l"])
            .output()
            .ok()?;

        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines() {
            if line.contains("IOPlatformSerialNumber") {
                if let Some(start) = line.find('"') {
                    let rest = &line[start + 1..];
                    if let Some(end) = rest.find('"') {
                        return Some(rest[..end].to_string());
                    }
                }
            }
        }
    }

    None
}

fn generate_fallback_id() -> String {
    // Use hostname + username as fallback
    let hostname = hostname::get()
        .map(|h| h.to_string_lossy().to_string())
        .unwrap_or_else(|_| "unknown".to_string());

    let username = whoami::username();

    format!("{}@{}", username, hostname)
}

fn hash_string(input: &str) -> String {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};

    let mut hasher = DefaultHasher::new();
    input.hash(&mut hasher);
    let hash1 = hasher.finish();

    // Do a second pass with different seed
    let reversed: String = input.chars().rev().collect();
    let mut hasher2 = DefaultHasher::new();
    reversed.hash(&mut hasher2);
    let hash2 = hasher2.finish();

    format!("{:016x}{:016x}", hash1, hash2)
}

File: client/src-wails/services/hwid.go
package services

import (
	"fmt"
	"hash/fnv"
	"os"
	"os/exec"
	"os/user"
	"runtime"
	"strings"
	"sync"
)

var (
	hwid     string
	hwidOnce sync.Once
)

func GetHwid() string {
	hwidOnce.Do(func() {
		hwid = generateHwid()
	})
	return hwid
}

func generateHwid() string {
	var components []string
	if cpuId := getCpuId(); cpuId != "" {
		components = append(components, cpuId)
	}
	if diskSerial := getDiskSerial(); diskSerial != "" {
		components = append(components, diskSerial)
	}
	if mac := getMacAddress(); mac != "" {
		components = append(components, mac)
	}
	if mbSerial := getMotherboardSerial(); mbSerial != "" {
		components = append(components, mbSerial)
	}
	if len(components) == 0 {
		components = append(components, generateFallbackId())
	}
	combined := strings.Join(components, "|")
	return hashString(combined)
}

func getCpuId() string {
	switch runtime.GOOS {
	case "windows":
		return runWmicCommand("cpu", "get", "ProcessorId")
	case "linux":
		content, err := os.ReadFile("/proc/cpuinfo")
		if err != nil {
			return ""
		}
		for _, line := range strings.Split(string(content), "\n") {
			if strings.HasPrefix(line, "Serial") || strings.Contains(line, "model name") {
				parts := strings.SplitN(line, ":", 2)
				if len(parts) == 2 {
					return strings.TrimSpace(parts[1])
				}
			}
		}
	case "darwin":
		out, err := exec.Command("sysctl", "-n", "machdep.cpu.brand_string").Output()
		if err == nil {
			return strings.TrimSpace(string(out))
		}
	}
	return ""
}

func getDiskSerial() string {
	switch runtime.GOOS {
	case "windows":
		return runWmicCommand("diskdrive", "get", "SerialNumber")
	case "linux":
		entries, err := os.ReadDir("/dev/disk/by-id")
		if err != nil {
			return ""
		}
		for _, entry := range entries {
			name := entry.Name()
			if strings.HasPrefix(name, "ata-") || strings.HasPrefix(name, "nvme-") {
				return name
			}
		}
	case "darwin":
		out, err := exec.Command("system_profiler", "SPSerialATADataType").Output()
		if err == nil {
			for _, line := range strings.Split(string(out), "\n") {
				if strings.Contains(line, "Serial Number") {
					parts := strings.SplitN(line, ":", 2)
					if len(parts) == 2 {
						return strings.TrimSpace(parts[1])
					}
				}
			}
		}
	}
	return ""
}

func getMacAddress() string {
	switch runtime.GOOS {
	case "windows":
		out, err := exec.Command("getmac", "/fo", "csv", "/nh").Output()
		if err != nil {
			return ""
		}
		for _, line := range strings.Split(string(out), "\n") {
			parts := strings.Split(line, ",")
			if len(parts) > 0 {
				mac := strings.Trim(strings.TrimSpace(parts[0]), "\"")
				if mac != "" && strings.Contains(mac, "-") {
					return mac
				}
			}
		}
	case "linux":
		entries, err := os.ReadDir("/sys/class/net")
		if err != nil {
			return ""
		}
		for _, entry := range entries {
			name := entry.Name()
			if name == "lo" {
				continue
			}
			addrPath := fmt.Sprintf("/sys/class/net/%s/address", name)
			content, err := os.ReadFile(addrPath)
			if err == nil {
				mac := strings.TrimSpace(string(content))
				if mac != "" && mac != "00:00:00:00:00:00" {
					return mac
				}
			}
		}
	case "darwin":
		out, err := exec.Command("ifconfig", "en0").Output()
		if err == nil {
			for _, line := range strings.Split(string(out), "\n") {
				if strings.Contains(line, "ether") {
					fields := strings.Fields(line)
					if len(fields) >= 2 {
						return fields[1]
					}
				}
			}
		}
	}
	return ""
}

func getMotherboardSerial() string {
	switch runtime.GOOS {
	case "windows":
		result := runWmicCommand("baseboard", "get", "SerialNumber")
		if result != "" && result != "To be filled by O.E.M." {
			return result
		}
	case "linux":
		content, err := os.ReadFile("/sys/class/dmi/id/board_serial")
		if err == nil {
			return strings.TrimSpace(string(content))
		}
	case "darwin":
		out, err := exec.Command("ioreg", "-l").Output()
		if err == nil {
			for _, line := range strings.Split(string(out), "\n") {
				if strings.Contains(line, "IOPlatformSerialNumber") {
					start := strings.Index(line, "\"")
					if start != -1 {
						rest := line[start+1:]
						end := strings.Index(rest, "\"")
						if end != -1 {
							return rest[:end]
						}
					}
				}
			}
		}
	}
	return ""
}

func runWmicCommand(args ...string) string {
	cmd := exec.Command("wmic", args...)
	cmd.SysProcAttr = getHideWindowAttr()
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	lines := strings.Split(string(out), "\n")
	for i, line := range lines {
		if i == 0 {
			continue
		}
		trimmed := strings.TrimSpace(line)
		if trimmed != "" {
			return trimmed
		}
	}
	return ""
}

func generateFallbackId() string {
	hostname, _ := os.Hostname()
	if hostname == "" {
		hostname = "unknown"
	}
	username := "unknown"
	if u, err := user.Current(); err == nil {
		username = u.Username
	}
	return fmt.Sprintf("%s@%s", username, hostname)
}

func hashString(input string) string {
	h1 := fnv.New64a()
	h1.Write([]byte(input))
	hash1 := h1.Sum64()
	reversed := reverseString(input)
	h2 := fnv.New64a()
	h2.Write([]byte(reversed))
	hash2 := h2.Sum64()
	return fmt.Sprintf("%016x%016x", hash1, hash2)
}

func reverseString(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}
